criu/apparmor.c

```diff
@@ -350,6 +350,112 @@ int collect_aa_namespace(char *profile)
 	return -1;
 }
 
+#define NEXT_AA_TOKEN(pos)					\
+	while (*pos) {						\
+		if (*pos == '/' &&				\
+				*(pos+1) && *(pos+1) == '/' &&	\
+				*(pos+2) && *(pos+2) == '&') {	\
+			pos += 3;				\
+			break;					\
+		}						\
+		if (*pos == ':' &&				\
+				*(pos+1) && *(pos+1) == '/' &&	\
+				*(pos+2) && *(pos+2) == '/') {	\
+			pos += 3;				\
+			break;					\
+		}						\
+		pos++;						\
+	}
+
+static int write_aa_policy(AaNamespace *ns, char *path, int offset, char *rewrite)
+{
+	int i, my_offset, ret;
+	char *rewrite_pos = rewrite, namespace[PATH_MAX];
+
+	if (!rewrite) {
+		strncpy(namespace, ns->name, sizeof(namespace) - 1);
+	} else {
+		NEXT_AA_TOKEN(rewrite_pos);
+
+		switch(*rewrite_pos) {
+		case ':': {
+			char tmp, *end;
+
+			end = strchr(rewrite_pos+1, ':');
+			if (!end) {
+				pr_err("invalid namespace %s\n", rewrite_pos);
+				return -1;
+			}
+
+			tmp = *end;
+			*end = 0;
+			strlcpy(namespace, rewrite_pos+1, sizeof(namespace));
+			*end = tmp;
+
+			break;
+		} default:
+			strlcpy(namespace, ns->name, sizeof(namespace));
+			for (i = 0; i < ns->n_policies; i++) {
+				if (strcmp(ns->policies[i]->name, rewrite_pos))
+					pr_warn("binary rewriting of apparmor policies not supported right now, not renaming %s to %s\n", ns->policies[i]->name, rewrite_pos);
+			}
+		}
+	}
+
+	my_offset = snprintf(path+offset, PATH_MAX-offset, "/namespaces/%s", ns->name);
+	if (my_offset < 0 || my_offset >= PATH_MAX-offset) {
+		pr_err("snprintf'd too many characters\n");
+		return -1;
+	}
+
+	if (mkdir(path, 0755) < 0 && errno != EEXIST) {
+		pr_perror("failed to create namespace %s", path);
+		goto fail;
+	}
+
+	for (i = 0; i < ns->n_namespaces; i++) {
+		if (write_aa_policy(ns, path, offset + my_offset, rewrite_pos) < 0)
+			goto fail;
+	}
+
+	ret = snprintf(path+offset+my_offset, sizeof(path)-offset-my_offset, "/.replace");
+	if (ret < 0 || ret >= sizeof(path)-offset-my_offset) {
+		pr_err("snprintf failed\n");
+		goto fail;
+	}
+
+	for (i = 0; i < ns->n_policies; i++) {
+		AaPolicy *p = ns->policies[i];
+		void *data = p->blob.data;
+		int fd, n;
+		off_t len = p->blob.len;
+
+		fd = open(path, O_WRONLY);
+		if (fd < 0) {
+			pr_perror("couldn't open apparmor load file %s", path);
+			goto fail;
+		}
+
+
+		n = write(fd, data, len);
+		close(fd);
+
+		if (n != len) {
+			pr_perror("write AA policy %s in %s failed", p->name, namespace);
+			goto fail;
+		}
+	}
+
+	return 0;
+
+fail:
+	path[offset + my_offset] = 0;
+ rmdir(path);
```

Comment:
Could you please adjust the indentation on this line?

Reply:
Fixed.